packages/storage-plus/src/indexes.rs

```diff
@@ -38,97 +37,183 @@ where
+        idx_fn: fn(&T, Vec<u8>) -> K,
+        pk_namespace: &'a str,
+        idx_namespace: &'a str,
+    ) -> Self {
         MultiIndex {
             index: idx_fn,
-            pk_map: Map::new(pk_namespace),
+            idx_namespace: idx_namespace.as_bytes(),
             idx_map: Map::new(idx_namespace),
+            pk_namespace: pk_namespace.as_bytes(),
         }
     }
 }
 
-impl<'a, T> Index<T> for MultiIndex<'a, T>
+fn deserialize_multi_kv<T: DeserializeOwned>(
+    store: &dyn Storage,
+    pk_namespace: &[u8],
+    kv: KV,
+) -> StdResult<KV<T>> {
+    let (key, pk_len) = kv;
+
+    // Deserialize pk_len
+    let pk_len = from_slice::<u32>(pk_len.as_slice())?;
+
+    // Recover pk from last part of k
+    let offset = key.len() - pk_len as usize;
+    let pk = &key[offset..];
+
+    let full_key = namespaces_with_key(&[pk_namespace], pk);
+
+    let v = store
+        .get(&full_key)
+        .ok_or_else(|| StdError::generic_err("pk not found"))?;
+    let v = from_slice::<T>(&v)?;
+
+    Ok((pk.into(), v))
+}
+
+impl<'a, K, T> Index<T> for MultiIndex<'a, K, T>
 where
     T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a>,
 {
     fn save(&self, store: &mut dyn Storage, pk: &[u8], data: &T) -> StdResult<()> {
-        let idx = (self.index)(data);
-        self.idx_map.save(store, (&idx, &pk), &MARKER)
+        let idx = (self.index)(data, pk.to_vec());
+        self.idx_map.save(store, idx, &(pk.len() as u32))
     }
 
     fn remove(&self, store: &mut dyn Storage, pk: &[u8], old_data: &T) -> StdResult<()> {
-        let idx = (self.index)(old_data);
-        self.idx_map.remove(store, (&idx, &pk));
+        let idx = (self.index)(old_data, pk.to_vec());
+        self.idx_map.remove(store, idx);
         Ok(())
     }
 }
 
-impl<'a, T> MultiIndex<'a, T>
+impl<'a, K, T> MultiIndex<'a, K, T>
 where
     T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a>,
 {
-    pub fn pks<'c>(
-        &self,
-        store: &'c dyn Storage,
-        idx: &[u8],
-        min: Option<Bound>,
-        max: Option<Bound>,
-        order: Order,
-    ) -> Box<dyn Iterator<Item = Vec<u8>> + 'c> {
-        let prefix = self.idx_map.prefix(idx);
-        let mapped = range_with_prefix(store, &prefix, min, max, order).map(|(k, _)| k);
-        Box::new(mapped)
+    pub fn prefix(&self, p: K::Prefix) -> Prefix<T> {
+        Prefix::with_deserialization_function(
+            self.idx_namespace,
+            &p.prefix(),
+            self.pk_namespace,
+            deserialize_multi_kv,
+        )
     }
 
-    /// returns all items that match this secondary index, always by pk Ascending
-    pub fn items<'c>(
-        &'c self,
+    pub fn sub_prefix(&self, p: K::SubPrefix) -> Prefix<T> {
+        Prefix::with_deserialization_function(
+            self.idx_namespace,
+            &p.prefix(),
+            self.pk_namespace,
+            deserialize_multi_kv,
+        )
+    }
+
+    // FIXME?: Move to Prefix<T> for ergonomics
```

Comment:
Is this comment outdated? You use K::Prefix

Or you mean to move the method? If so, that can wait til later.

Reply:
I mean that, yes (move the method), will clarify the comment.